File: python/nvec.py
import array

"""
    Vector Library for Machine Learning:
    Author: Stewart Charles
    Version: 1.0
"""
class MlVec(object):
    # Initialize MlVec with Data
    def __init__(self, data, type='d'):
        self.data = array.array(type,data)
        self.dim = len(self.data)
        self.type = type

        # Mapping Function
    def map(self,fn,rhs = None):
        ans = MlVec.of_size(self.dim,self.type)
        if rhs is None:
            for i in range(0,self.dim):
                ans[i] = fn(self[i])
        else:
            if not isinstance(rhs,MlVec):
                for i in range(0,self.dim):
                    ans[i] = fn(self[i],rhs)
            else:
                for i in range(0,self.dim):
                    ans[i] = fn(self[i],rhs[i])
        return ans

    # Fold Function
    def fold(self,fn,rhs = None, accum = 0):
        if rhs is None:
            for i in range(0, self.dim):
                accum = fn(self[i],accum)
        else:
            if not isinstance(rhs,MlVec):
                for i in range(0, self.dim):
                    accum = fn(self[i],rhs,accum)
            else:
                for i in range(0, self.dim):
                    accum = fn(self[i],rhs[i],accum)
        return accum

    # Create an empty MlVec initialized to Zero
    @classmethod
    def of_size(cls,size,type = 'd'):
        return cls([0] * size, type)

    # Operator Overload for [] get
    def __getitem__(self,pos):
        return self.data[pos]

    # Operator Overload for [] set
    def __setitem__(self,pos,item):
        self.data[pos] = item

    # -- Arithmatic Operators --

    def __add__(self,rhs):
        return self.map(lambda a,b: a + b, rhs)

    def __sub__(self,rhs):
        return self.map(lambda a,b: a - b, rhs)

    def __mul__(self,rhs):
        return self.map(lambda a,b: a * b, rhs)

    def __div__(self,rhs):
        if not isinstance(rhs, MlVec):
            return self * (1.0 / rhs)
        else:
            return self.map(lambda a,b: a / b, rhs)

    def dot(self,rhs):
        return self.fold(lambda a,b,c: c + a * b, rhs)

    # Overload the Power operator for MlVec to Performa Dot Product
    # if RHS is an MlVec, otherwise performs pow element-wise using rhs as a scalar
    def __pow__ (self, rhs):
        if isinstance(rhs,MlVec):
            return self.dot(rhs)
        else:
            return self.map(lambda a,b: a ** b, rhs)

    # Equality Operator Overload
    def __eq__ (self,rhs):
        sum = self.fold(lambda x,y,z: z + (x == y), rhs)
        return sum == self.dim

    def __ne__ (self,rhs):
        return not (self == rhs)

# Testing Functions
a = MlVec([1,1,1,1])
b = MlVec([2,3,4,5])
test = a + b
assert(test == MlVec([3,4,5,6]))

test = a * 2
assert(test == MlVec([2,2,2,2]))

test_num = (a * 2) ** (a * 2)
assert(test_num == 16)

test = a * b + a * b
assert(test == MlVec([4,6,8,10]))

test = a - a
assert(test == MlVec([0,0,0,0]))

test = b / 2
assert(test == MlVec([1,1.5,2,2.5]))

test = b / b
assert(test == a)
print("Tests Passed")

File: swift/nvec.swift
//
//  nvec.swift
//  MachineLearning
//
//  Created by Stewart Charles on 12/11/16.
//  Copyright © 2016 Stewart Charles. All rights reserved.
//

import Foundation
class MlVec {
  
  var data : [Float64]
  
  init(_ d:[Float64]) {
    data = d
  }
  
  init(_ d: Float64 ...) {
    data = d
  }
  
  func len() -> Int {
    return data.count
  }
  
  func map(fn:(Float64)->Float64)-> MlVec {
    var arr = [Float64] (repeating: 0.0, count: len())
    for i in 0..<len() {
      arr[i] = fn(data[i])
    }
    return MlVec(arr)
  }
  
  func map(rhs:MlVec, fn:(Float64,Float64)->Float64)->MlVec {
    var arr = [Float64] (repeating: 0.0, count: len())
    for i in 0..<len() {
      arr[i] = fn(data[i], rhs[i])
    }
    return MlVec(arr)
  }
  
  func fold(fn:(Float64,Float64)->Float64,accum:Float64)->Float64 {
    var ans = accum
    for i in 0..<len() {
      ans = fn(data[i],ans)
    }
    return ans
  }
  
  func fold(rhs:MlVec,fn:(Float64,Float64,Float64)->Float64,accum:Float64) -> Float64 {
    var ans = accum
    for i in 0..<len() {
      ans = fn(data[i],rhs[i],ans)
    }
    return ans
  }
  
  func add(rhs:MlVec)->MlVec {
    return map(rhs:rhs,fn:{(a: Float64, b : Float64)->Float64 in a + b})
  }
  
  func sub(rhs:MlVec)->MlVec {
    return map(rhs:rhs,fn:{(a: Float64, b : Float64)->Float64 in a - b})
  }
  
  func mul(rhs:MlVec)->MlVec {
    return map(rhs:rhs,fn:{(a: Float64, b : Float64)->Float64 in a * b})
  }
  
  func div(rhs:MlVec)->MlVec {
    return map(rhs:rhs,fn:{(a: Float64, b : Float64)->Float64 in a / b})
  }
  
  func add(rhs:Float64)->MlVec {
    return map(fn:{(a: Float64)->Float64 in a + rhs})
  }
  
  func sub(rhs:Float64)->MlVec {
    return map(fn:{(a: Float64)->Float64 in a - rhs})
  }
  
  func mul(rhs:Float64)->MlVec {
    return map(fn:{(a: Float64)->Float64 in a * rhs})
  }
  
  func div(rhs:Float64)->MlVec {
    return mul(rhs:rhs)
  }
  
  func dot(rhs:MlVec)->Float64 {
    return fold(rhs:rhs, fn:{(a: Float64, b: Float64, c:Float64)->Float64 in a * b + c}, accum:0.0)
  }
  
  static func +(_ lhs:MlVec,_ rhs:MlVec)->MlVec {
    return lhs.add(rhs: rhs)
  }
  
  static func -(_ lhs:MlVec,_ rhs:MlVec)->MlVec {
    return lhs.sub(rhs: rhs)
  }
  
  static func /(_ lhs:MlVec,_ rhs:MlVec)->MlVec {
    return lhs.div(rhs: rhs)
  }
  
  static func *(_ lhs:MlVec,_ rhs:MlVec)->MlVec {
    return lhs.mul(rhs: rhs)
  }
  
  static func +(_ lhs:MlVec,_ rhs:Float64)->MlVec {
    return lhs.add(rhs: rhs)
  }
  
  static func -(_ lhs:MlVec,_ rhs:Float64)->MlVec {
    return lhs.sub(rhs: rhs)
  }
  
  static func *(_ lhs:MlVec,_ rhs:Float64)->MlVec {
    return lhs.mul(rhs: rhs)
  }
  
  static func /(_ lhs:MlVec,_ rhs:Float64)->MlVec {
    return lhs.div(rhs: rhs)
  }
  
  static func *(_ lhs:Float64,_ rhs:MlVec)->MlVec {
    return rhs.mul(rhs: lhs)
  }
  
  static func ==(_ lhs:MlVec,_ rhs: MlVec)->Bool {
    let size = Float64(lhs.len())
    let test = lhs.fold(rhs: rhs,
                        fn: {(a:Float64,b:Float64,c:Float64)->Float64 in
                         let test = a == b
                         return (c + (test ? 1.0 : 0.0))},
                        accum: 0.0)
    return size == test
  }
  
  static func !=(_ lhs:MlVec,_ rhs: MlVec)->Bool {
    return !(lhs == rhs)
  }
  
  subscript(_ index:Int)->Float64 {
    get { return data[index] }
    set { data[index] = newValue }
  }
  
}
